pkg/gather/collect/collect.go

```diff
@@ -0,0 +1,488 @@
+}
+
+func (c *Collector) collectPod(ctx context.Context, u *unstructured.Unstructured, resourceInfo *ResourceInfo) error {
+	pod := &corev1.Pod{}
+	err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, pod)
+	if err != nil {
+		return fmt.Errorf("can't convert secret from unstructured: %w", err)
+	}
+
+	err = c.writeResource(ctx, pod, resourceInfo)
+	if err != nil {
+		return err
+	}
+
+	logsDir := filepath.Join(c.getResourceDir(pod, resourceInfo), pod.GetName())
+
+	err = os.MkdirAll(logsDir, 0770)
+	if err != nil {
+		return fmt.Errorf("can't create logs dir %q: %w", logsDir, err)
+	}
+
+	for _, container := range pod.Spec.InitContainers {
+		err = c.collectContainerLogs(ctx, logsDir, pod, &container)
+		if err != nil {
+			return fmt.Errorf("can't collect logs for init container %q in pod %q: %w", container.Name, naming.ObjRef(pod), err)
+		}
+	}
+
+	for _, container := range pod.Spec.Containers {
+		err = c.collectContainerLogs(ctx, logsDir, pod, &container)
+		if err != nil {
+			return fmt.Errorf("can't collect logs for container %q in pod %q: %w", container.Name, naming.ObjRef(pod), err)
+		}
+	}
+
+	return nil
+}
+
+func (c *Collector) DiscoverResources(ctx context.Context, filter discovery.ResourcePredicateFunc) ([]*ResourceInfo, error) {
+	all, err := c.discoveryClient.ServerPreferredResources()
+	if err != nil {
+		return nil, fmt.Errorf("can't discover resources: %w", err)
+	}
+
+	rls := discovery.FilteredBy(filter, all)
+
+	// There should be at least one resource per group, likely more.
+	resourceInfos := make([]*ResourceInfo, 0, len(rls))
+	for _, rl := range rls {
+		gv, err := schema.ParseGroupVersion(rl.GroupVersion)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, r := range rl.APIResources {
+			var scope meta.RESTScope
+			if r.Namespaced {
+				scope = meta.RESTScopeNamespace
+			} else {
+				scope = meta.RESTScopeRoot
+			}
+			resourceInfos = append(resourceInfos, &ResourceInfo{
+				Scope:    scope,
+				Resource: gv.WithResource(r.Name),
+			})
+		}
+	}
+
+	return resourceInfos, nil
+}
+
+func ReplaceIsometricresourceInfosIfPresent(resourceInfos []*ResourceInfo) ([]*ResourceInfo, error) {
+	// Replacements are order dependent and should start from the oldest API.
+	replacements := []struct {
+		old schema.GroupVersionResource
+		new schema.GroupVersionResource
+	}{
+		{
+			old: schema.GroupVersionResource{
+				Group:    "",
+				Version:  "v1",
+				Resource: "events",
+			},
+			new: schema.GroupVersionResource{
+				Group:    "events.k8s.io",
+				Version:  "v1",
+				Resource: "events",
+			},
+		},
+	}
+
+	resourceInfosMap := make(map[schema.GroupVersionResource]*ResourceInfo, len(resourceInfos))
+	for _, m := range resourceInfos {
+		resourceInfosMap[m.Resource] = m
+	}
+
+	for _, replacement := range replacements {
+		_, found := resourceInfosMap[replacement.new]
+		if found {
+			delete(resourceInfosMap, replacement.old)
+		}
+	}
+
+	return maps.Values(resourceInfosMap), nil
```

Comment:
We have `helpers.GetMapValues`. I'd say if you're fine with using experimental package then let's get rid of our Map helpers.

Reply:
Yup, I guess it can go as a part of a PR cleaning the helpers out. @zimnx wdyt about the above?

Reply:
I would restrain from using experimental packages as they are meant to be removed eventually which adds additional work when it's moved to stdlib.

Reply:
> I would restrain from using experimental packages as they are meant to be removed eventually 

I can rebase to pickup that function (since it is already there) but I don't think I agree with this statement. Any function in k8s.io/apimachinery or `k8s.io/client-go` can go away or be renamed on any version bump as well.

In the end the risk is primarily a rename when it moves to the stdlib which is quite easy to update. Compared to that custom functions risk implementation bugs and we lack unit tests too often. (Say looking at `GetMapValues` compared too https://github.com/golang/exp/blob/79cabaa/maps/maps_test.go#L35)

Reply:
done

Reply:
The difference between functions in third party dependencies and golang experimental package is expermiental contains stuff that is supposed to be removed without major version bump. Third party dependencies should follow backward compatibility guarantees and only remove exported stuff when major version of a module is bumped. 
So if we have same semantic available under our package I would prefer to use it compared to experimental package, but it's not something i would insist on so it wouldn't block my approval.